crates/exex/exex/src/manager.rs

```diff
@@ -303,83 +312,152 @@ impl ExExManager {
     }
 }
 
-impl Future for ExExManager {
+impl<P> ExExManager<P>
+where
+    P: HeaderProvider,
+{
+    /// Finalizes the WAL according to the passed finalized header.
+    ///
+    /// This function checks if all ExExes are on the canonical chain and finalizes the WAL if
+    /// necessary.
+    fn finalize_wal(&self, finalized_header: SealedHeader) -> eyre::Result<()> {
+        debug!(header = ?finalized_header.num_hash(), "Received finalized header");
+
+        // Check if all ExExes are on the canonical chain
+        let exex_finished_heights = self
+            .exex_handles
+            .iter()
+            // Get ExEx ID and hash of the finished height for each ExEx
+            .map(|exex_handle| {
+                (&exex_handle.id, exex_handle.finished_height.map(|block| block.hash))
+            })
+            // Deduplicate all hashes
+            .unique_by(|(_, hash)| *hash)
+            // Check if hashes are canonical
+            .map(|(exex_id, hash)| {
+                hash.map_or(Ok((exex_id, hash, false)), |hash| {
+                    self.provider
+                        .is_known(&hash)
+                        // Save the ExEx ID, hash of the finished height, and whether the hash
+                        // is canonical
+                        .map(|is_canonical| (exex_id, Some(hash), is_canonical))
+                })
+            })
+            // We collect here to be able to log the unfinalized ExExes below
+            .collect::<Result<Vec<_>, _>>()?;
+        if exex_finished_heights.iter().all(|(_, _, is_canonical)| *is_canonical) {
+            // If there is a finalized header and all ExExs are on the canonical chain, finalize
+            // the WAL with the new finalized header
+            self.wal.finalize(finalized_header.num_hash())?;
+        } else {
+            let unfinalized_exexes = exex_finished_heights
+                .into_iter()
+                .filter_map(|(exex_id, hash, is_canonical)| {
+                    is_canonical.not().then_some((exex_id, hash))
+                })
+                .collect::<Vec<_>>();
```

Comment:
what I meant was this is redundant because only used for debug trace and we now have access to 

https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.format